charts/TimeBounds.ts

```diff
@@ -0,0 +1,129 @@
+ * cannot enforce this in TypeScript.
+ */
+export type Time = number
+
+/**
+ * An unbounded value (±Infinity) or a concrete point in time (year or date).
+ */
+export type TimeBound = number
+
+/**
+ * The two special TimeBound values: unbounded left & unbounded right.
+ */
+export enum TimeBoundValue {
+    unboundedLeft = -Infinity,
+    unboundedRight = Infinity
+}
+
+export enum TimeBoundValueStr {
+    unboundedLeft = "earliest",
+    unboundedRight = "latest"
+}
+
+export function isTime(time: TimeBound | Time): time is Time {
+    return !isUnbounded(time)
+}
+
+export function isUnbounded(time: TimeBound): time is TimeBoundValue {
+    return isUnboundedLeft(time) || isUnboundedRight(time)
+}
+
+export function isUnboundedLeft(time: TimeBound): time is TimeBoundValue {
+    return time === TimeBoundValue.unboundedLeft
+}
+
+export function isUnboundedRight(time: TimeBound): time is TimeBoundValue {
+    return time === TimeBoundValue.unboundedRight
+}
+
+export function formatTimeBound(time: TimeBound): string {
+    if (isUnboundedLeft(time)) {
+        return TimeBoundValueStr.unboundedLeft
+    }
+    if (isUnboundedRight(time)) {
+        return TimeBoundValueStr.unboundedRight
+    }
+    return `${time}`
+}
+
+export function parseTimeBound(
+    str: string,
+    defaultTo: TimeBoundValue
+): TimeBound {
+    if (str === TimeBoundValueStr.unboundedLeft) {
+        return TimeBoundValue.unboundedLeft
+    }
+    if (str === TimeBoundValueStr.unboundedRight) {
+        return TimeBoundValue.unboundedRight
+    }
+    const time = parseIntOrUndefined(str)
+    return time !== undefined ? time : defaultTo
+}
+
+// Use this to not repeat logic
+function fromJSON(
+    value: TimeBound | string | undefined,
+    defaultValue: TimeBound
+) {
+    if (isString(value)) return parseTimeBound(value, defaultValue)
+    if (value === undefined) return defaultValue
+    return value
+}
+
+function toJSON(bound: TimeBound | undefined): string | number | undefined {
+    if (bound === undefined) {
+        return undefined
+    }
+    if (isUnboundedLeft(bound)) {
+        return TimeBoundValueStr.unboundedLeft
+    }
+    if (isUnboundedRight(bound)) {
+        return TimeBoundValueStr.unboundedRight
+    }
+    return bound
+}
+
+export function minTimeFromJSON(
+    minTime: TimeBound | string | undefined
+): TimeBound {
+    return fromJSON(minTime, TimeBoundValue.unboundedLeft)
+}
+
+export function maxTimeFromJSON(
+    maxTime: TimeBound | string | undefined
+): TimeBound {
+    return fromJSON(maxTime, TimeBoundValue.unboundedRight)
+}
+
+export const minTimeToJSON = toJSON
+export const maxTimeToJSON = toJSON
+
+export function getTimeFromTimeRange(
+    [minTime, maxTime]: [Time, Time],
+    bound: TimeBound
+): Time {
+    if (isUnboundedLeft(bound)) return minTime
+    if (isUnboundedRight(bound)) return maxTime
+    return Math.min(maxTime, Math.max(minTime, bound))
+}
+
+export function getTimeFromTimes(
```

Comment:
I find the name here... not so great

Reply:
Haha nope! It's hard to name these things. The differences are quite subtle, I didn't manage to come up with a good one – let me know if you have suggestions

I opted for docstrings to clear up some parts, but that's definitely not ideal.

Reply:
Not a direct suggestion, but `findClosestYear` is a way better name and is similar to what this does. So, maybe `getClosestTime`?

Reply:
Good point, it's using `findClosestYear` so it should have `closest` in the name. Now changed, thanks!